### PR TITLE
Change Resource Queue locking and rewrite the way resource locks are managed

### PR DESCRIPTION
**Pull request by atris**
Change the locking logic for Resource Queues and prevent deadlocks occurring due to them. This also makes COST based resource checking optional which has to be set using resource_usecost GUC.

Thanks @lpetrov-pivotal for design ideas, code reviews and design reviews and other help!

### CODE REVIEW

**Comment by ivannovick** [comment]
@hlinnaka @eurekaka this code is in sensitive area, can you please help with a good review of this?

**Comment by ivannovick** [comment]
@lpetrov-pivotal thx for your work on this

File: src/backend/utils/resscheduler/resscheduler.c [hunk 1]
```diff
@@ -1085,3 +1086,91 @@ ResHandleUtilityStmt(Portal portal, Node *stmt)
 		portal->status = PORTAL_ACTIVE;
 	}
 }
+
+/*
+ * Pre lock the current slot in Resource Queue
```
Comment (hlinnaka) [inline src/backend/utils/resscheduler/resscheduler.c]: What does "pre-locking" a slot mean? (before what, does the "pre-" refer to?)

**Comment by hlinnaka** [comment]
I'm missing the big picture here. What is the problem we're trying to solve, and how does this PR solve it?

On a more detailed level, it would make review much easier if you'd clean up all the whitespace issues in a separate patch, and squash all the actual changes together.

Is adding the new GUC needed to avoid the deadlocks? If not, please split that off to a separate PR so we can discuss that separately.

**Comment by hlinnaka** [comment]
A brief intro to how resource queue locks and slots work in general would be nice. I've never really been able to figure that out.

**Comment by atris** [comment]
The main problem here is that using Resource Queues with shared database object locks leads to a high potential of deadlocks. The root cause of this problem is that there is no fixed locking order between RQ slot locks and database shared object locks.Further diving deep into the problem, the RQ slots locks are acquired at Portal creation time through various paths (Portal creation, SPI call etc). This leads to a circular wait problem where a normal UPDATE statement may acquire Exclusive lock on relation in parsing stage.
The PR solves this problem by assuring that RQ slot locks are always acquired prior to acquiring any database shared object locks. The RQ slot lock held is released at transaction finish processing. An explicit transaction's queries share the same RQ slot lock. The primary blocker to this approach is the COST based checking in Resource Queues, which can be done after planning stage. After PM discussion, the COST based checking is semi deprecated and is now controlled with resource_usecost GUC. If resource_usecost GUC is ON, the old behaviour is applicable but with the GUC off, the new behaviour takes place. This ensures that we can proceed with pre locking the RQ slot lock safely and can be assured that the quota used per query remains the same.

**Comment by atris** [comment]
@hlinnaka Fixed. Please see and comment.

**Comment by ivannovick** [comment]
@eurekaka can you comment on this PR?

**Comment by hlinnaka** [comment]
The patch needs a lot more comments, in particular, high-level comments that explain how resource queue locks work in general. Why are they acquired at the places that they are? Why are they released where they are (and not where other locks are released)? What does "pre-locking" mean?

I wonder if exec_simple_query() is the right place for the new code. For starters, that only applies to simple queries, which misses prepared statements completely.

I think the new GUC should be added as a separate commit, it seems like an independent part from the rest of the patch.

File: src/backend/utils/resscheduler/resscheduler.c (hunk 1 above)
Comment (hlinnaka) [inline src/backend/utils/resscheduler/resscheduler.c]: Thanks, that new comment helps a lot. Now that I understand this better, I have more questions (which should probably be answered in code comments):

1. Why does cost-based checking work differently?

2. I don't like the term "pre-locking". It's still not clear what the "pre-" refers to. Before what does it happen. I'd call it just "locking". This functions hould probably be called just ResLockAcquire(), like for normal locks, the main entry point is called LockAcquire().  I see that there is a function called that already, though, but it seems like a more low-level function. Perhaps rename ResLockAcquire() to something else? Not sure, but in general, I think the naming needs some further thought.

3. Please make it clear that the rule is: "always acquire resource lock first, before any other locks". The comment almost says that, when it talks about deadlocks, but I think it should be said more explicitly.

Comment (hlinnaka) [inline src/backend/utils/resscheduler/resscheduler.c]: regarding 3., there's some explanation of deadlocks between resource locks and normal locks in the beginning of the README. That needs to be updated.

File: src/backend/tcop/postgres.c [hunk 2]
```diff
@@ -1999,6 +2088,87 @@ exec_parse_message(const char *query_string,	/* string to execute */
 		Snapshot        mySnapshot = NULL;
 		int			i;
 
+		memset(&incData, 0x00, sizeof(incData));
+		switch (nodeTag(parsetree))
```
Comment (hlinnaka) [inline src/backend/tcop/postgres.c]: I'd suggest putting this code to a separate function, so that it doesn't need to be duplicated between exec_simple_query() and exec_parse_message()

Comment (hlinnaka) [inline src/backend/tcop/postgres.c]: If we do this checking only in exec_parse_message(), I think that means that executing an already-prepared statement skips the locking. I'd suggest writing a test suite for this, exercising different scenarios, using libpq.

File: src/backend/access/transam/xact.c [hunk 3]
```diff
@@ -3428,6 +3428,9 @@ CommitTransaction(void)
 	/* Close any open regular cursors */
 	AtCommit_Portals();
 
+	if (!ResourceQueueUseCost)
```
Comment (hlinnaka) [inline src/backend/access/transam/xact.c]: There's no harm in calling ResLockPreUnlock() always, right? I'd leave out the "if()" condition, to match the style of other similar cleanup-calls here, and do the "ResourceQueueUseCost" check inside ResLockPreUnlock().

File: src/backend/utils/resscheduler/resscheduler.c (hunk 1 above)
Comment (atris) [inline src/backend/utils/resscheduler/resscheduler.c]: Fixed,thanks

File: src/backend/tcop/postgres.c (hunk 2 above)
Comment (atris) [inline src/backend/tcop/postgres.c]: Fixed,thanks.

File: src/backend/access/transam/xact.c (hunk 3 above)
Comment (atris) [inline src/backend/access/transam/xact.c]: Fixed,thanks

File: src/backend/tcop/postgres.c [hunk 4]
```diff
@@ -2670,6 +2846,20 @@ exec_execute_message(const char *portal_name, int64 max_rows)
 		portalParams = portal->portalParams;
 	}
 
+	current_stmt = FetchPreparedStatement(prepStmtName, false);
+
+	/*
+	 * When the prepared statement was first stored in the structure, ResourceQueueUseCost
+	 * Queue Slot lock was acquired. This is for case when a stored prepared statement is
+	 * executed */
+	if (current_stmt)
```
Comment (hlinnaka) [inline src/backend/tcop/postgres.c]: The comment says that the slot lock was acquired earlier already. So why is this needed at all? Or, if we do it here, why bother to do it earlier?

Comment (atris) [inline src/backend/tcop/postgres.c]: This is for the case when prepared statement is executed (not parsed and executed). When a prepared statement is parsed, lock is taken in parse stage since we should protect underlying relation when processing. For the case when we execute a stored prepared statement, we need to take the lock during execution stage, hence this check

**Comment by hlinnaka** [comment]
This doesn't apply cleanly anymore, thanks to the 8.3 merge commit. Can you rebase over current master, please?

**Comment by yydzero** [comment]
New resource management design is under development, close this PR for now.
